modules/picard_qc/1.0/picard_qc.smk

```diff
@@ -0,0 +1,323 @@
+        CFG["threads"]["rnaseq_metrics"]
+    resources:
+        mem_mb = CFG["mem_mb"]["rnaseq_metrics"]
+    shell:
+        op.as_one_line("""
+        picard -Xmx{resources.mem_mb}m CollectRnaSeqMetrics {params.opts} 
+        I={input.bam} O={output.metrics} 
+        R={input.fasta} 
+        STRAND_SPECIFICITY={params.strand} 
+        REF_FLAT={input.refFlat}
+        RIBOSOMAL_INTERVALS={input.rrna_int}
+        CHART_OUTPUT={output.metrics}.pdf 
+        > {log.stdout} 2> {log.stderr}
+        """)
+
+# genome metrics
+rule _picard_qc_wgs_metrics:
+    input:
+        bam = rules._picard_qc_input_bam.output.bam,
+        fasta = reference_files("genomes/{genome_build}/genome_fasta/genome.fa")
+    output:
+        metrics = CFG["dirs"]["metrics"] + "{seq_type}--{genome_build}/{sample_id}/wgs_metrics"
+    log:
+        stdout = CFG["logs"]["metrics"] + "{seq_type}--{genome_build}/{sample_id}/wgs_metrics.stdout.log",
+        stderr = CFG["logs"]["metrics"] + "{seq_type}--{genome_build}/{sample_id}/wgs_metrics.stderr.log"
+    params:
+        opts = CFG["options"]["wgs_metrics"],
+    conda:
+        CFG["conda_envs"]["picard"]
+    threads:
+        CFG["threads"]["wgs_metrics"]
+    resources:
+        mem_mb = CFG["mem_mb"]["wgs_metrics"]
+    shell:
+        op.as_one_line("""
+        picard -Xmx{resources.mem_mb}m CollectWgsMetrics {params.opts} 
+        I={input.bam} O={output.metrics} 
+        R={input.fasta} 
+        > {log.stdout} 2> {log.stderr}
+        """)
+
+
+def _get_sample_metrics(metrics_dir):
+    DIR = metrics_dir
+    def _get_sample_metrics_custom(wildcards):
+        CFG = config["lcr-modules"]["picard_qc"]
+        sample = op.filter_samples(CFG["samples"], seq_type=wildcards.seq_type)
+        return expand("{dir}{seq_type}--{genome_build}/{sample_id}/{metrics}", dir = DIR, sample_id = list(sample["sample_id"]), **wildcards)
+    return _get_sample_metrics_custom
+
+
+rule _picard_qc_merge_metrics:
+    input: 
+        _get_sample_metrics(metrics_dir = CFG["dirs"]["metrics"])
+    output: 
+        metrics = CFG["dirs"]["merged_metrics"] + "{seq_type}--{genome_build}/all.{metrics}.txt"
+    run:
+        samples = input
+        with open(output[0], "w") as out:
+            for i in range(0,len(samples)):
+                s_id = samples[i].split("/")[-2]
+                with open(samples[i], "r") as f:
+                    data = [l for l in f.readlines() if not (l.startswith('#') or l == '\n')]
+                    if i == 0:
+                        header = "SAMPLEID\t" + "\t".join(data[0].split("\t")[0:])
+                        out.write(header)
+                    if wildcards.metrics == "alignment_summary_metrics":
+                        line = s_id + "\t" + "\t".join(data[3].split("\t")[0:])
+                    else:
+                        line = s_id + "\t" + "\t".join(data[1].split("\t")[0:])
+                    out.write(line)
+                    f.close()
+        out.close()
```

Comment:
A lot is going on here. Could you briefly explain what's happening here?

Reply:
I wonder if the `run` portion can be captured as a bash script to avoid loading stuff into Python.

Reply:
This is to create a merged metrics file with all the samples from a seq_type. The reason we wrote it in python was because it was originally a shell script but at a certain number of samples it fails because it reached a max character limit. The original shell script looks something like 
'grep -v '#' {input[0]} | sed '/^$/d' | head -n 1 | cut -f 2- > {output}.header &&
        for file in {input}; do
            echo 'Processing file:' ${{file}};
            grep -v '#' ${{file}} | sed '/^$/d' | sed -n 2p | cut -f 2- >> {output}.tmp;
        done &&
        echo {params.samples} | sed 's/ /\\n/g' | paste - {output}.tmp > {output} &&
        echo -e 'sampleID' | paste - {output}.header | cat - {output} > {output}.tmp &&
        mv {output}.tmp {output} && rm {output}.header`

Reply:
I think this is a good way to get around the issues with character limits. I wonder if it would be worth discussing an even more general strategy that could be implemented as python function, possibly in oncopipe? Something that takes in a list of files to merge, and has options for which lines to skip, which lines to keep, etc.? Maybe the 3 of us can discuss further.

Reply:
hmm ya that could be useful, are there any other cases where we would want to merge a file that looks like this? Because I think for vcf files there are tools out there that does it for you

Reply:
My other ideas would have been mutation calls (although we maybe should just use a vcf/maf merge tool as you mentioned) or even copy number calls. But I guess right now the only immediately obvious application is the QC files.

Reply:
I'm  not sure if it would work for multiple file formats but if we need to merge other stuff in the future we can think about converting it into an oncopipe function instead